File: android-native/app/src/main/java/flexr/social/app/data/repository/WithdrawalRepository.kt
package flexr.social.app.data.repository

import flexr.social.app.core.network.apiCall
import flexr.social.app.data.remote.FlexrApi
import flexr.social.app.data.remote.dto.WithdrawalRequestDto
import flexr.social.app.domain.model.WithdrawalAck
import javax.inject.Inject
import javax.inject.Singleton

/** Online-Rücktrittsfunktion (§ 13a FAGG). Entspricht backend/app/routers/withdrawal.py. */
@Singleton
class WithdrawalRepository @Inject constructor(
    private val api: FlexrApi,
) {

    suspend fun declare(
        name: String,
        email: String,
        contractReference: String?,
        message: String?,
        requestId: String,
        language: String,
    ): WithdrawalAck {
        val ack = apiCall {
            api.declareWithdrawal(
                WithdrawalRequestDto(
                    name = name.trim(),
                    email = email.trim(),
                    contractReference = contractReference?.trim()?.ifBlank { null },
                    message = message?.trim()?.ifBlank { null },
                    confirmed = true,
                    requestId = requestId,
                    language = language,
                ),
            )
        }
        return WithdrawalAck(
            reference = ack.reference,
            declarationText = ack.declarationText,
            confirmationSent = ack.confirmationSent,
            message = ack.message,
        )
    }
}

File: ios/FLEXR/Data/Repository/WithdrawalRepository.swift
import Foundation

/// Online-Rücktrittsfunktion (§ 13a FAGG). Entspricht backend/app/routers/withdrawal.py.
@MainActor
final class WithdrawalRepository {

    private let api: FlexrAPI

    init(api: FlexrAPI) {
        self.api = api
    }

    func declare(
        name: String,
        email: String,
        contractReference: String?,
        message: String?,
        requestId: String,
        language: String
    ) async throws -> WithdrawalAck {
        let trimmedContractReference = contractReference?.trimmingCharacters(in: .whitespacesAndNewlines)
        let trimmedMessage = message?.trimmingCharacters(in: .whitespacesAndNewlines)
        let ack = try await api.declareWithdrawal(
            WithdrawalRequestDTO(
                name: name.trimmingCharacters(in: .whitespacesAndNewlines),
                email: email.trimmingCharacters(in: .whitespacesAndNewlines),
                contractReference: (trimmedContractReference?.isEmpty ?? true) ? nil : trimmedContractReference,
                message: (trimmedMessage?.isEmpty ?? true) ? nil : trimmedMessage,
                confirmed: true,
                requestId: requestId,
                language: language
            )
        )
        return WithdrawalAck(
            reference: ack.reference,
            declarationText: ack.declarationText,
            confirmationSent: ack.confirmationSent,
            message: ack.message
        )
    }
}
